### PR TITLE
VectorToImageTransform conversion to estimator/transformer

### LINKED ISSUE #2269: Lockdown Microsoft.ML.ImageAnalytics public surface
We ought to internalize as much of that assembly to those things we want to be public.

### PR DESCRIPTION
**Pull request by yaeldekel**
This is the remaining work for the public API of Microsoft.ML.ImageAnalytics. 
Fixes #2269 .

### CODE REVIEW

File: src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs [hunk 1]
```diff
@@ -226,195 +533,63 @@ internal void Save(ModelSaveContext ctx)
...
             {
-                _exes[i] = new ColInfoEx(ctx);
-                _types[i] = new ImageType(_exes[i].Height, _exes[i].Width);
-            }
-            Metadata.Seal();
-        }
+                if (!inputSchema.TryFindColumn(colInfo.InputColumnName, out var col))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.InputColumnName);
+                if (col.Kind != SchemaShape.Column.VectorKind.Vector || (col.ItemType != NumberDataViewType.Single && col.ItemType != NumberDataViewType.Double && col.ItemType != NumberDataViewType.Byte))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.InputColumnName, new ImageType().ToString(), col.GetTypeString());
```
Comment (TomFinley) [inline src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs]: >new ImageType().ToString() [](start = 107, length = 26)

This is not correct. You're saying the expected input type is an image type, yet, you are testing whether it is a known-size vector of float, double, or byte above.

File: src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs [hunk 2]
```diff
@@ -139,17 +401,27 @@ private sealed class ColInfoEx
             public readonly Single Scale;
             public readonly bool Interleave;
 
-            public bool Alpha { get { return (Colors & ColorBits.Alpha) != 0; } }
-            public bool Red { get { return (Colors & ColorBits.Red) != 0; } }
-            public bool Green { get { return (Colors & ColorBits.Green) != 0; } }
-            public bool Blue { get { return (Colors & ColorBits.Blue) != 0; } }
+            public bool Alpha { get { return (Colors & ImagePixelExtractingEstimator.ColorBits.Alpha) != 0; } }
```
Comment (TomFinley) [inline src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs]: >public bool Alpha { get { return (Colors & ImagePixelExtractingEstimator.ColorBits.Alpha) != 0; } } [](start = 12, length = 99)

Not a big deal, but if you wanted to bring this transplanted code up to "modern standards" by using `=>` style syntax to make this a bit less crazy taht would be fine with me.

File: src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs [hunk 3]
```diff
@@ -114,23 +125,274 @@ public class Arguments : TransformInputBase
...
+                disposer = null;
+                var sourceType = InputSchema[ColMapNewToOld[iinfo]].Type;
+                var sourceItemType = sourceType.GetItemType();
+                if (sourceItemType == NumberDataViewType.Single || sourceItemType == NumberDataViewType.Double)
+                    return GetterFromType<float>(input, iinfo, ex, needScale);
+                else
+                    if (sourceItemType == NumberDataViewType.Byte)
+                    return GetterFromType<byte>(input, iinfo, ex, false);
+                else
+                    throw Contracts.Except("We only support float or byte arrays");
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs]: >We only support float or byte arrays [](start = 44, length = 36)

From estimator: `"known-size vector of type float, double or byte`
Double get lost.

File: test/Microsoft.ML.Tests/ImagesTests.cs [hunk 4]
```diff
@@ -201,13 +201,13 @@ public void TestBackAndForthConversionWithAlphaInterleave()
             var cropped = new ImageResizingTransformer(env, "ImageCropped", imageWidth, imageHeight, "ImageReal").Transform(images);
 
             var pixels = new ImagePixelExtractingTransformer(env, "ImagePixels", "ImageCropped", ImagePixelExtractingEstimator.ColorBits.All, true, 2f / 255, 127.5f).Transform(cropped);
-            IDataView backToBitmaps = new VectorToImageTransform(env, new VectorToImageTransform.Arguments()
+            IDataView backToBitmaps = VectorToImageConvertingTransformer.Create(env, new VectorToImageConvertingTransformer.Options()
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/ImagesTests.cs]: >VectorToImageConvertingTransformer.Create [](start = 38, length = 41)

Please don't do that.

First it's a different style comparing to all other transformers.

Second Create method is IDataTransform which we want to get rid of.

File: src/Microsoft.ML.ImageAnalytics/ExtensionsCatalog.cs [hunk 5]
```diff
@@ -133,5 +133,30 @@ public static ImagePixelExtractingEstimator ExtractPixels(this TransformsCatalog
...
+        /// <param name="colors"> Specifies which <see cref="ImagePixelExtractingEstimator.ColorBits"/> are in the input pixel vectors. The order of colors is: Alpha, Red, Green Blue.</param>
+        /// <param name="interleave">Whether the pixels are interleaved, meaning whether they are in `ARGB ARGB` order, or separated in the planar form, where the colors are specified one by one
+        /// alpha, red, green, blue for all the pixels of the image. </param>
+        /// <param name="scale">The values are scaled by this value before being converted to pixels.</param>
+        /// <param name="offset">The offset is subtracted (before scaling) before converting the values to pixels.</param>
+        public static VectorToImageConvertingEstimator ConvertToImage(this TransformsCatalog catalog, int height, int width, string outputColumnName, string inputColumnName = null,
+            ImagePixelExtractingEstimator.ColorBits colors = ImagePixelExtractingEstimator.ColorBits.Rgb, bool interleave = false, float scale = 1, float offset = 0)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.ImageAnalytics/ExtensionsCatalog.cs]: >1 [](start = 145, length = 1)

would be nice to have `Defaults` class

File: src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs [hunk 6]
```diff
@@ -108,29 +119,289 @@ public class Arguments : TransformInputBase
...
+        internal const string UserName = "Vector To Image Transform";
+        internal const string LoaderSignature = "VectorToImageConverter";
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "VECTOIMG",
+                //verWrittenCur: 0x00010001, // Initial
+                //verWrittenCur: 0x00010002, // Swith from OpenCV to Bitmap
+                verWrittenCur: 0x00010003, // don't serialize sizeof(Single)
+                verReadableCur: 0x00010003,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs]: >0x00010003 [](start = 32, length = 10)

You can always have check for model version and read one Int32 from stream if it's old one.

I have separate PR where I bump version for ImagePixelExtractor (I introduce colors order). So I plan do the same for VectorToImage and bump version again.
I can get rid of single and add order in same time.
Does it sound good for you?
If you still want to get rid of single in this iteration, I would suggest to keep old models readable.

File: src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs [hunk 7]
```diff
@@ -108,29 +119,289 @@ public class Arguments : TransformInputBase
...
+
+        private readonly VectorToImageConvertingEstimator.ColumnInfo[] _columns;
+
         /// <summary>
-        /// Which color channels are extracted. Note that these values are serialized so should not be modified.
+        /// The columns passed to this <see cref="ITransformer"/>.
         /// </summary>
-        [Flags]
-        private enum ColorBits : byte
+        public IReadOnlyCollection<VectorToImageConvertingEstimator.ColumnInfo> Columns => _columns.AsReadOnly();
```
Comment (zeahmed) [inline src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs]: >public [](start = 8, length = 6)

internal???

**Pull request merged**